### PR TITLE
expression: implement vectorized evaluation for' builtinCastStringAsDecimalSig'

### PR DESCRIPTION
**Pull request by ZhaoQian888**
PCP  #12102
### What problem does this PR solve? 
PCP #12102
implement vectorized evaluation for' builtinCastStringAsDecimalSig'
### What is changed and how it works?

```
BenchmarkVectorizedBuiltinCastFunc/builtinCastStringAsDecimalSig-VecBuiltinFunc-8                  20570             58290 ns/op               0 B/op          0 allocs/op

BenchmarkVectorizedBuiltinCastFunc/builtinCastStringAsDecimalSig-NonVecBuiltinFunc-8                8769            114479 ns/op           49152 B/op       1024 allocs/op
```
### Check List 

Tests 

 - Unit test
 - Integration test

### CODE REVIEW

File: expression/builtin_cast_vec.go [hunk 1]
```diff
@@ -1272,11 +1273,65 @@ func (b *builtinCastStringAsRealSig) vecEvalReal(input *chunk.Chunk, result *chu
...
+	for i := 0; i < n; i++ {
+		if result.IsNull(i) {
+			continue
+		}
+		tempdecimal := new(types.MyDecimal)
+		if !(b.inUnion && mysql.HasUnsignedFlag(b.tp.Flag) && res[i].IsNegative()) {
+			te := buf.GetString(i)
+			err = sc.HandleTruncate(tempdecimal.FromString([]byte(te)))
+			if err != nil {
+				fmt.Println("I am here3")
```
Comment (wshwsh12) [inline expression/builtin_cast_vec.go]: ditto.

File: expression/builtin_cast_vec.go [hunk 2]
```diff
@@ -1272,11 +1272,47 @@ func (b *builtinCastStringAsRealSig) vecEvalReal(input *chunk.Chunk, result *chu
...
-	return false
+	return true
 }
 
 func (b *builtinCastStringAsDecimalSig) vecEvalDecimal(input *chunk.Chunk, result *chunk.Column) error {
-	return errors.Errorf("not implemented")
+	n := input.NumRows()
+	if IsBinaryLiteral(b.args[0]) {
+		if err := b.args[0].VecEvalDecimal(b.ctx, input, result); err != nil {
+			return err
```
Comment (Reminiscent) [inline expression/builtin_cast_vec.go]: Should we ``` return b.args[0].VecEvalDecimal(b.ctx, input, result)``` directly?  @qw4990  @wshwsh12 
Please add some test cases to cover such situations.

Comment (wshwsh12) [reply expression/builtin_cast_vec.go]: Yes.

**Comment by qw4990** [comment]
@ZhaoQian888 Friendly ping, please address comments.

File: expression/builtin_cast_vec.go (hunk 2 above)
Comment (ZhaoQian888) [reply expression/builtin_cast_vec.go]: > @ZhaoQian888 Friendly ping, please address comments.

I'm busy these days ,I will fix it when I have enough time

**Comment by ZhaoQian888** [comment]
``` golang
func (c *castAsDecimalFunctionClass) getFunction(ctx sessionctx.Context, args []Expression) (sig builtinFunc, err error) {
	if err := c.verifyArgs(args); err != nil {
		return nil, err
	}
	bf := newBaseBuiltinCastFunc(newBaseBuiltinFunc(ctx, args), ctx.Value(inUnionCastContext) != nil)
	bf.tp = c.tp
	if IsBinaryLiteral(args[0]) {
		sig = &builtinCastDecimalAsDecimalSig{bf}
		sig.setPbCode(tipb.ScalarFuncSig_CastDecimalAsDecimal)
		return sig, nil
	}
```
In this Function ,I find it's no need to check whether   b.args[0]. IsBinaryLiteral() 
because in  `getFunction` it will go to` builtinCastDecimalAsDecimalSig`
therefore I think we should delete these code

**Comment by ZhaoQian888** [comment]
I  add   test cases to cover such situations.  return b.args[0].VecEvalDecimal(b.ctx, input)  @Reminiscent

File: expression/builtin_cast_vec.go [hunk 3]
```diff
@@ -1407,11 +1407,47 @@ func (b *builtinCastStringAsRealSig) vecEvalReal(input *chunk.Chunk, result *chu
...
+		}
+		tempdecimal := new(types.MyDecimal)
+		if !(b.inUnion && mysql.HasUnsignedFlag(b.tp.Flag) && res[i].IsNegative()) {
+			te := buf.GetString(i)
+			err = sc.HandleTruncate(tempdecimal.FromString([]byte(te)))
+			if err != nil {
+				return err
+			}
+		}
+		_, err = types.ProduceDecWithSpecifiedTp(tempdecimal, b.tp, sc)
```
Comment (qw4990) [inline expression/builtin_cast_vec.go]:
```suggestion
		tempdecimal, err = types.ProduceDecWithSpecifiedTp(tempdecimal, b.tp, sc)
```

**Comment by qw4990** [comment]
@ZhaoQian888 Sorry for my late reply.
Please resolve conflicts and address [this comment](https://github.com/pingcap/tidb/pull/13566#discussion_r354674671).

**Comment by ZhaoQian888** [comment]
@qw

**Comment by ZhaoQian888** [comment]
there is someone has already done this function,that' the reason there is confiction @qw4990  I will close it

**Pull request closed without merging**